integration_tests/data_utils.py

```diff
@@ -0,0 +1,199 @@
+
+Hypothesis is used for most of the data generation, which will hopefully improve the rigour
+of our tests when we can roll it into more tests.
+"""
+import string
+from datetime import datetime, timedelta
+from pathlib import Path
+
+import numpy as np
+import rasterio
+import yaml
+from hypothesis.strategies import sampled_from, datetimes, composite, floats, lists, text, uuids
+
+from datacube.utils.geometry import CRS, point
+
+DATE_FORMAT = '%Y-%m-%d %H:%M:%S'
+
+# Bounds retrieved from http://www.epsg-registry.org
+# Code, lat_bounds, lon_bounds
+EPSG_CODES = [('EPSG:32755', (-80, 0), (144, 150))]
+crses = sampled_from(EPSG_CODES)
+REASONABLE_DATES = datetimes(min_value=datetime(1980, 1, 1), max_value=datetime(2030, 1, 1))
+
+
+@composite
+def bounds(draw, bounds):
+    return draw(
+        lists(
+            floats(*bounds), min_size=2, max_size=2
+        ).map(sorted).filter(lambda x: x[0] < x[1]))
+
+
+@composite
+def bboxes(draw, lat_bounds=(-90, 90), lon_bounds=(-180, 180)):
+    """
+    https://tools.ietf.org/html/rfc7946#section-5
+    """
+    # Use 3 dim positions even if we only need 2
+    min_lat, max_lat = draw(bounds(lat_bounds))
+    min_lon, max_lon = draw(bounds(lon_bounds))
+
+    return min_lon, min_lat, max_lon, max_lat
+
+
+@composite
+def extents(draw, lat_bounds, lon_bounds):
+    center_dt = draw(REASONABLE_DATES)
+
+    time_range = timedelta(seconds=12)
+    min_lon, min_lat, max_lon, max_lat = draw(bboxes(lat_bounds, lon_bounds))
+    return {
+        'center_dt': center_dt.strftime(DATE_FORMAT),
+        'from_dt': (center_dt - time_range).strftime(DATE_FORMAT),
+        'to_dt': (center_dt + time_range).strftime(DATE_FORMAT),
+        'coord': {
+            'll': {'lat': min_lat, 'lon': min_lon},
+            'lr': {'lat': min_lat, 'lon': max_lon},
+            'ul': {'lat': max_lat, 'lon': min_lon},
+            'ur': {'lat': max_lat, 'lon': max_lon},
+        }
+    }
+
+
+def _extent_point_projector(crs):
+    crs = CRS(crs)
+
+    def reproject_point(pos):
+        pos = point(pos['lon'], pos['lat'], CRS('EPSG:4326'))
+        coords = pos.to_crs(crs).coords[0]
+        return {'x': coords[0], 'y': coords[1]}
+
+    return reproject_point
+
+
+def extent_to_grid_spatial(extent, crs):
+    """Convert an extent in WGS84 to a grid spatial in the supplied CRS"""
+    reprojector = _extent_point_projector(crs)
+    return {
+        'projection': {
+            'geo_ref_points': {
+                corner: reprojector(pos)
+                for corner, pos in extent['coord'].items()
+            },
+            'spatial_reference': crs
+        }
+    }
+
+
+@composite
+def acquisition_details(draw):
+    return {
+        'aos': draw(REASONABLE_DATES),
+        'groundstation': {
+            'code': draw(text(alphabet=string.ascii_letters, min_size=3, max_size=5))
+        },
+        'los': draw(REASONABLE_DATES)
+    }
+
+
+@composite
+def scene_datasets(draw):
```

Comment:
How does this `@composite` + `draw` magic works? Who defines draw? where?

Reply:
It's magic from [hypothesis](https://hypothesis.readthedocs.io/en/latest/) (automated testing framework, similar to QuickCheck for Haskell). I'm abusing it a bit here to generate single random datasets, whereas it's supposed to be used to run each test tens of times with randomly generated data that's likely to expose bugs.

I want to move in that direction, and start using some of these on our tests, but want to be careful about not blowing the run time for the test suite out of the galaxy.